src/Client.sol

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {DataCapTypes} from "filecoin-solidity/contracts/v0.8/types/DataCapTypes.sol";
+import {DataCapAPI} from "filecoin-solidity/contracts/v0.8/DataCapAPI.sol";
+import {Errors} from "./libs/Errors.sol";
+import {CBORDecoder} from "filecoin-solidity/contracts/v0.8/utils/CborDecode.sol";
+import {IClient} from "./interfaces/IClient.sol";
+import {Multicall} from "@openzeppelin/contracts/utils/Multicall.sol";
+import {BigInts} from "filecoin-solidity/contracts/v0.8/utils/BigInts.sol";
+
+contract Client is Initializable, IClient, Multicall {
+    address public manager;
+
+    mapping(address client => uint256 allowance) public allowances;
+    mapping(address client => mapping(uint64 storageProvider => bool allowedSP) allowedSPs) public clientSPs;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address manager_) public initializer {
+        manager = manager_;
+    }
+
+    /**
+     * @notice This function transfers DataCap tokens from the client to the storage provider
+     * @dev This function can only be called by the client
+     * @param params The parameters for the transfer
+     * @dev Reverts with InsufficientAllowance if caller doesn't have sufficient allowance
+     * @dev Reverts with InvalidAmount when parsing amount from BigInt to uint256 failed
+     */
+    function transfer(DataCapTypes.TransferParams calldata params) external {
+        (uint256 parsedAmount, bool failed) = BigInts.toUint256(params.amount);
+        if (failed) revert Errors.InvalidAmount();
+        if (allowances[msg.sender] < parsedAmount) revert Errors.InsufficientAllowance();
+        uint64[] memory providers = _deserializeAllocationRequests(params.operator_data, 0);
+        _matchAllowedSPs(providers);
+        allowances[msg.sender] -= parsedAmount;
+        /// @custom:oz-upgrades-unsafe-allow-reachable delegatecall
+        DataCapAPI.transfer(params);
+        emit DatacapAllocated(msg.sender, params.to, params.amount);
+    }
+
+    /**
+     * @dev This function sets the list of allowed storage providers for a specific client
+     * @param client The address of the client for whom the allowed storage providers are being set
+     * @param allowedSPs_ The list of allowed storage providers
+     */
+    function addAllowedSPsForClient(address client, uint64[] calldata allowedSPs_) external onlyManager {
+        for (uint256 i = 0; i < allowedSPs_.length; i++) {
+            clientSPs[client][allowedSPs_[i]] = true;
+        }
+        emit SPsChanged(allowedSPs_);
```

Comment:
We need to change this event.
1. Right now the same event is used for adding and removal, so we don't know what was actually done with these SPs (so split into two events)
2. It lacks `client` in topics, so we don't know who this change is for.